### PR TITLE
feat: add optional FireRed streaming VAD

### PR DESCRIPTION
**Pull request by tavallaie**
Ticket: #235

## Problem

Issue 235 asked for FireRedVAD as an optional detector next to Silero. Silero stays the default.

## Changes

- `--vad firered` with extra `fireredvad` and `--vad_firered_model_dir`.
- `FireRedVadIterator` uses FireRed `is_speech_start` / `is_speech_end`. It does not subclass Silero `VADIterator`.
- Incoming audio stays float32 in `[-1, 1]`. FireRed gets int16-scale windows and a 400/160 overlap tail.
- Active speech is counted in 10 ms hops. Trailing hangover silence is not counted.
- A 20 s FireRed split can emit end then start in one 512-sample chunk. The new start stays buffered. The ended hop count is kept for `VADHandler`.
- Setup docs use `hf download`. `huggingface-cli` is gone in `huggingface_hub` 1.0.

## Behaviour after the change

- `--vad silero` is unchanged.
- `--vad firered` fails at setup if the extra or `--vad_firered_model_dir` is missing.
- No JSON key change.

## Verification

```
uv run mypy src/
Success: no issues found in 104 source files

uv run ruff check src/ tests/
All checks passed!

uv run ruff format --check src/ tests/
171 files already formatted

uv run pytest tests/test_firered_vad_iterator.py tests/test_cli_defaults.py tests/test_backend_registry.py -q
115 passed
```

## Follow-up

Parakeet Unified STT is #553. Nemotron streaming STT is #554.

### CODE REVIEW

File: src/speech_to_speech/VAD/firered_vad_iterator.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, streamer: FireRedStreamer) -> None:
+        self.streamer = streamer
+
+    def reset_states(self) -> None:
+        self.streamer.reset()
+
+    def __call__(self, x: torch.Tensor, sampling_rate: int) -> torch.Tensor:
+        samples = x.detach().cpu().contiguous().view(-1).numpy().astype(np.float32, copy=False)
+        # FireRed fbank trains on int16 PCM. VADIterator feeds Silero-scale [-1, 1].
+        results = self.streamer.detect_chunk(samples * 32768.0)
```
Comment (andimarafioti) [inline src/speech_to_speech/VAD/firered_vad_iterator.py]: Passing each disjoint 512-sample chunk directly to `detect_chunk` loses the overlap FireRed needs. Its frontend uses 400-sample windows with a 160-sample step and restarts feature extraction on each call. For 3.2 seconds of audio, this produces 100 frames instead of 318; with the real model and its English sample, speech detection starts about 130 ms later and ends about 350 ms later than continuous framing. Please retain the unconsumed waveform tail, feed complete windows at the required step, and clear that buffer on reset. A test covering successive 512-sample chunks would catch this—the current mock always returns one score. [FireRed's streaming example](https://github.com/FireRedTeam/FireRedVAD/blob/main/fireredvad/bin/stream_vad.py) shows the required overlap.

File: src/speech_to_speech/VAD/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+Streaming FireRedVAD has no published benchmark yet. Silero remains the default.
+
+## FireRedVAD
+
+Install the optional extra and download Stream-VAD weights:
+
+```bash
+pip install "speech-to-speech[fireredvad]"
+huggingface-cli download FireRedTeam/FireRedVAD --local-dir ./pretrained_models/FireRedVAD
```
Comment (andimarafioti) [inline src/speech_to_speech/VAD/README.md]: `huggingface-cli` was removed in `huggingface_hub` 1.0, so this command fails on a fresh installation with current dependencies before the user can download the required weights. Please change it to `hf download FireRedTeam/FireRedVAD --local-dir ./pretrained_models/FireRedVAD`. [Migration docs](https://huggingface.co/docs/huggingface_hub/main/concepts/migration#cli-cache-commands).

**Review by andimarafioti — CHANGES_REQUESTED** [review]
There are a couple of issues to fix before merging: the FireRed adapter loses analysis frames between audio chunks, and the setup instructions use a removed command. Details inline.

File: src/speech_to_speech/VAD/README.md (hunk 2 above)
Comment (tavallaie) [reply src/speech_to_speech/VAD/README.md]: Addressed in 0e1faf9: the VAD README now uses `hf download FireRedTeam/FireRedVAD --local-dir ./pretrained_models/FireRedVAD`.

File: src/speech_to_speech/VAD/firered_vad_iterator.py (hunk 1 above)
Comment (tavallaie) [reply src/speech_to_speech/VAD/firered_vad_iterator.py]: Addressed in 0e1faf9: FireRedProbModel now concatenates the unconsumed hop-160 tail before each detect_chunk call, feeds complete 400-sample windows, and clears that tail on reset. A test of 100 successive 512-sample chunks now yields 318 frames.

File: src/speech_to_speech/VAD/firered_vad_iterator.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+                self._pre_speech_buffer.clear()
+                self._pre_speech_samples = 0
+                self.buffer.append(x)
+                chunk_in_buffer = True
+                self.active_speech_samples = self._num_samples(x)
+                self.last_utterance_active_speech_samples = 0
+            if frame.is_speech_end and self.triggered:
+                if not chunk_in_buffer:
+                    self.buffer.append(x)
+                return self._end_utterance()
```
Comment (andimarafioti) [inline src/speech_to_speech/VAD/firered_vad_iterator.py]: This return skips any remaining frame events in the chunk. With FireRed's default 20-second limit, frame 2000 emits a speech end and frame 2001 emits a new start; both can arrive in the same 512-sample chunk. The start gets discarded, leaving the iterator inactive while the user continues speaking. With the real FireRed postprocessor and 22 seconds of continuous speech scores, only the first 20.032 seconds were emitted and the rest was unbuffered. Please process the remaining events and preserve the subsequent start/audio before returning, with a regression test for this boundary.

File: src/speech_to_speech/VAD/firered_vad_iterator.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        # detect_chunk restarts fbank, so feed complete 400-sample windows at a 160-sample hop.
+        n_frames = (len(audio) - _FIRERED_WINDOW_SAMPLES) // _FIRERED_HOP_SAMPLES + 1
+        chunk_end = _FIRERED_WINDOW_SAMPLES + (n_frames - 1) * _FIRERED_HOP_SAMPLES
+        results = self.streamer.detect_chunk(audio[:chunk_end])
+        self._tail = audio[n_frames * _FIRERED_HOP_SAMPLES :]
+        return results
+
+    def _append_chunk(self, x: torch.Tensor) -> None:
+        self.buffer.append(x)
+        self.active_speech_samples += self._num_samples(x)
```
Comment (andimarafioti) [inline src/speech_to_speech/VAD/firered_vad_iterator.py]: This counts every buffered chunk as active speech, including silence while FireRed waits to emit its end event. With a 500 ms silence threshold, 100 ms of speech scores becomes 480 ms of "active speech", passes the handler's 384 ms minimum, and emits an interrupting speech-start event followed by a final utterance. A brief sound can therefore interrupt the assistant during the silence that follows it. Please count speech-classified frame duration separately from buffered audio duration, and test that trailing silence cannot make a short sound pass the minimum.

**Review by andimarafioti — CHANGES_REQUESTED** [review]
The previous issues are fixed, but there are two issues in the new FireRed event handling: continuing speech can be dropped at the 20-second split, and trailing silence is counted as active speech. Details inline.

**Comment by andimarafioti** [comment]
I also think it would be simpler if this were a few different PRs for each thing you're trying to add :)

File: src/speech_to_speech/VAD/firered_vad_iterator.py (hunk 4 above)
Comment (tavallaie) [reply src/speech_to_speech/VAD/firered_vad_iterator.py]: Addressed in f79a576: active_speech_samples now counts FireRed speech hops (160 samples each), not trailing silence in the PCM buffer. A 10 ms speech hop plus hangover silence still reports 160 active samples.

File: src/speech_to_speech/VAD/firered_vad_iterator.py (hunk 3 above)
Comment (tavallaie) [reply src/speech_to_speech/VAD/firered_vad_iterator.py]: Addressed in f79a576 and 6f6bbc6: after is_speech_end the iterator keeps scanning the same 512-sample chunk. A following is_speech_start stays buffered and triggered. last_utterance_active_speech_samples keeps the ended hop count so VADHandler still sees the first utterance's active speech.

**Comment by tavallaie** [comment]
> I also think it would be simpler if this were a few different PRs for each thing you're trying to add :)

because I have it in single issue I did it in single PR. let me break it to few PR.

**Comment by tavallaie** [comment]
Split as requested. This PR is FireRed VAD only. Parakeet Unified and Nemotron STT are in https://github.com/huggingface/speech-to-speech/pull/553.

**Comment by andimarafioti** [comment]
Codex' still complaining:

3. **Blocking finding — medium severity, within scope.**

   **Speech before start confirmation is omitted from the duration count** at [[firered_vad_iterator.py:196–198](https://github.com/huggingface/speech-to-speech/pull/549/files)](codex://review?pr=https%3A%2F%2Fgithub.com%2Fhuggingface%2Fspeech-to-speech%2Fpull%2F549&path=src%2Fspeech_to_speech%2FVAD%2Ffirered_vad_iterator.py&line=197&side=right).

   FireRed requires eight consecutive speech frames before confirming a start. The iterator begins counting only on that confirmation, losing the preceding seven speech frames—70 ms.

   Reproducing this through FireRed’s actual postprocessor and the existing handler produced **430 ms of classified speech, counted as 360 ms**. The default 384 ms minimum then discarded it, producing no transcription or speech lifecycle events.

   Preserve the candidate speech-frame count and include it when the start is confirmed, without counting padding or hangover silence. Add a regression demonstrating that this short utterance reaches transcription.

**Comment by tavallaie** [comment]
@andimarafioti  Thanks for catching this. FireRed confirms start after eight speech frames, and we were only counting from that confirmation.

`4bc4688` keeps a candidate hop count while waiting for start and includes it when the start fires. 430 ms of speech is now 43 hops, not 36, so it still clears the 384 ms minimum. Pad and hangover silence stay out of the count.

**Review by andimarafioti — APPROVED** [review]
Good work, thank you! In the end codex kept on complaining so I just had it work on it instead of going back and forth, nothing big!
